### PR TITLE
Support private key encrypted with empty password

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

A user may use a private key which is encrypted with an empty password. Because of this we should only handle a null password in a special way.

Modifications:

Correctly handle private key that is encrypted with empty password.

Result:

Correctly support private key that is encrypted with empty password.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trustin PTAL...

File: handler/src/test/java/io/netty/handler/ssl/SslContextTest.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.ByteArrayInputStream;
+import java.security.PrivateKey;
+
+public class SslContextTest {
+
+    private static final String ENCRYPTED_EMPTY_PASSWORD = "-----BEGIN ENCRYPTED PRIVATE KEY-----\n" +
```
Comment (trustin) [inline handler/src/test/java/io/netty/handler/ssl/SslContextTest.java]: Perhaps you could move this into the resources directory?

Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SslContextTest.java]: as file ? Sure if you like. Not sure it matters though :)

**Comment by trustin** [comment]
I realized that `null` password and an empty password are different things. For example, a private key can be encrypted a really empty password, and it is not an unencrypted password.

Instead of assuming unencrypted private key for an empty password, could we try both cases (unencrypted and encrypted with an empty password) and choose the one that works? i.e.

```java
if (password == null || password.isEmpty()) {
    // try unencrypted. if failed, try encrypted with an empty password
} else {
    // try encrypted
}
```

**Comment by normanmaurer** [comment]
@trustin sure we could try both.. that said which exception message (and wrapped trace)  we want to throw if non of these work out ?

**Comment by trustin** [comment]
@normanmaurer Yeah, exactly.

**Comment by normanmaurer** [comment]
@trustin not sure what "Yeah, exactly" should tell me for my question ;)

**Comment by trustin** [comment]
@normanmaurer Ahh. I'm not sure. Maybe it doesn't matter a lot if we provide helpful exception message?

**Comment by normanmaurer** [comment]
@trustin actually I'm not sure anymore if you comment makes sense to first try one and then the other. Why should we do this if already told in our javadoc that the user should use `null` for not password. ?

**Comment by trustin** [comment]
@normanmaurer For better user experience? It's good without this PR, but we will get less reports about this. We could handle this case automatically while logging some warning to educate them, though. WDYT?

**Comment by normanmaurer** [comment]
@trustin PTAL again.

File: handler/src/main/java/io/netty/handler/ssl/SslContext.java [hunk 2]
```diff
@@ -933,24 +933,38 @@ private static PrivateKey getPrivateKeyFromByteBuffer(ByteBuf encodedKeyBuf, Str
         byte[] encodedKey = new byte[encodedKeyBuf.readableBytes()];
         encodedKeyBuf.readBytes(encodedKey).release();
 
-        PKCS8EncodedKeySpec encodedKeySpec = generateKeySpec(keyPassword == null ? null : keyPassword.toCharArray(),
-                                                             encodedKey);
+        try {
+            return generatePrivateKey(
+                    generateKeySpec(keyPassword == null ? null : keyPassword.toCharArray(), encodedKey));
+        } catch (IOException e) {
+            if (keyPassword != null && keyPassword.length() == 0) {
```
Comment (trustin) [inline handler/src/main/java/io/netty/handler/ssl/SslContext.java]: Could you also try with an empty password when keyPassword is `null`?

**Comment by trustin** [comment]
LGTM

**Comment by normanmaurer** [comment]
@nmittler PTAL

File: handler/src/main/java/io/netty/handler/ssl/OpenSslServerContext.java [hunk 3]
```diff
@@ -317,106 +317,27 @@ public OpenSslServerContext(
...
             X509Certificate[] trustCertChain, TrustManagerFactory trustManagerFactory,
             X509Certificate[] keyCertChain, PrivateKey key, String keyPassword, KeyManagerFactory keyManagerFactory,
             Iterable<String> ciphers, CipherSuiteFilter cipherFilter, ApplicationProtocolConfig apn,
             long sessionCacheSize, long sessionTimeout, ClientAuth clientAuth) throws SSLException {
+        this(trustCertChain, trustManagerFactory, keyCertChain, key, keyPassword, keyManagerFactory, ciphers,
+                cipherFilter, toNegotiator(apn), sessionCacheSize, sessionTimeout, clientAuth);
+    }
+
+    @SuppressWarnings("deprecation")
+    private OpenSslServerContext(
```
Comment (netkins) [inline handler/src/main/java/io/netty/handler/ssl/OpenSslServerContext.java]: MAJOR Constructor has 12 parameters, which is greater than 7 authorized. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS00107)

Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/OpenSslServerContext.java]: it's fine

**Comment by normanmaurer** [comment]
@Scottmitch PTAL

File: handler/src/main/java/io/netty/handler/ssl/OpenSslServerContext.java (hunk 3 above)
Comment (nmittler) [inline handler/src/main/java/io/netty/handler/ssl/OpenSslServerContext.java]: :)

**Comment by nmittler** [comment]
@normanmaurer LGTM ... a lot simpler :)

**Comment by Scottmitch** [comment]
lgtm

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (5c023976899d7b218dac91d7e8ff160545583dd4) and 4.0 (0a87c989937781b6e283a13e50e35466095bb3d0)

**Pull request closed without merging**